module/zfs/dsl_scan.c

```diff
@@ -630,6 +630,8 @@ dsl_scan_init(dsl_pool_t *dp, uint64_t txg)
 		zap_cursor_fini(&zc);
 	}
 
+	ddt_walk_init(spa, scn->scn_phys.scn_min_txg);
```

Comment:
The txg is here is the scan start point. If we have any log entries from before that txg, we need to flush the log out to up to that txg to ensure the ZAPs are correct.

If I've misunderstood, and `scn_min_txg` is not the scan start, then yes, maybe I need to change it! Regardless, I'll go back over all this shortly to make sure I understand it properly.

Reply:
Scan operates on data between `scn_min_txg` and `scn_max_txg`.  If you want to avoid having logs in that rage, then I think you'd want `ddl_first_txg > scn_max_txg`.

Reply:
Yep, confirmed my error now, and fixed. Along the way spotted a bug where we wouldn't clear force-flush until both logs were empty, which wasn't impossible when flushing aggressively but also wasn't great or needed. Code is much tidier now.